api/v1alpha1/rule_types.go

```diff
@@ -47,41 +61,137 @@ type RuleSpec struct {
+	// URL defines the target URL for incoming requests
 	URL string `json:"url"`
+	// StripPath replaces the provided path prefix when forwarding the requested URL to the upstream URL.
 	// +optional
 	StripPath *string `json:"stripPath,omitempty"`
+	// PreserveHost includes the host and port of the url value if set to false. If true, the host and port of the ORY Oathkeeper Proxy will be used instead.
 	// +optional
 	PreserveHost *bool `json:"preserveHost,omitempty"`
 }
 
+// Match defines the URL(s) that an access rule should match.
 type Match struct {
-	URL     string   `json:"url"`
+	// URL is the URL that should be matched. It supports regex templates.
+	URL string `json:"url"`
+	// Methods represent an array of HTTP methods (e.g. GET, POST, PUT, DELETE, ...)
 	Methods []string `json:"methods"`
 }
 
+// Authenticator represents a handler that authenticates provided credentials.
 type Authenticator struct {
 	*Handler `json:",inline"`
 }
 
+// Authorizer represents a handler that authorizes the subject ("user") from the previously validated credentials making the request.
 type Authorizer struct {
 	*Handler `json:",inline"`
 }
 
+// Mutator represents a handler that transforms the HTTP request before forwarding it.
 type Mutator struct {
 	*Handler `json:",inline"`
 }
 
+// Handler represents an Oathkeeper routine that operates on incoming requests. It is used to either validate a request (Authenticator, Authorizer) or modify it (Mutator).
 type Handler struct {
+	// Name is the name of a handler
 	Name string `json:"handler"`
+	// Config configures the handler. Configuration keys vary per handler.
 	// +kubebuilder:validation:Type=object
 	Config *runtime.RawExtension `json:"config,omitempty"`
 }
 
+// ToOathkeeperRules transforms a RuleList object into a JSON object digestible by Oathkeeper.
+func (rl RuleList) ToOathkeeperRules() ([]byte, error) {
+
+	rules := make([]*RuleJSON, len(rl.Items))
+
+	for i, _ := range rl.Items {
+		rules[i] = rl.Items[i].ToRuleJSON()
+	}
+
+	return json.MarshalIndent(rules, "", "  ")
+}
+
+// FilterNotValid filters out Rules which doesn't pass validation due to being not processed yet or due to negative result of validation. It returns a list of Rules which passed validation successfully.
+func (rl RuleList) FilterNotValid() RuleList {
+	rlCopy := rl
+	validRules := []Rule{}
+	for _, rule := range rl.Items {
+		if rule.Status.Validation != nil && rule.Status.Validation.Valid != nil && *rule.Status.Validation.Valid {
+			validRules = append(validRules, rule)
+		}
+	}
+	rlCopy.Items = validRules
+	return rlCopy
+}
+
+// ValidateWith uses provided validation configuration to check whether the rule have proper handlers set.
+func (r Rule) ValidateWith(config validation.Config) error {
+	// if a handler is not provided in any of the following fields, a default one will be used, so there is no need to check if it is valid
+	if r.Spec.Authenticators != nil {
+		for _, authenticator := range r.Spec.Authenticators {
```

Comment:
This will return on first found wrong Authenticator, what if there are mutliple?

Reply:
all invalid authenticators are now logged in cr